### PR TITLE
[NEW] Add drag-and-drop feature to move messages between folders

### PR DESCRIPTION
**Pull request by jacob-js**
## Pullrequest
Introduce drag-and-drop of messages between folders
### Issues

- [X] https://github.com/cypht-org/cypht/issues/216

### Demo
[drag-and-drop-demo.webm](https://github.com/cypht-org/cypht/assets/68013195/511710bc-a131-455f-907c-c27ddddb6b96)

### CODE REVIEW

File: modules/core/site.js [hunk 1]
```diff
@@ -1857,3 +1857,82 @@ function listControlsMenu() {
...
+        if (row.querySelector('.checkbox_cell input[type=checkbox]:checked')) {
+            selectedRows.push(row);
+        }
+    });
+    
+    if (selectedRows.length == 0) {
+        selectedRows.push(movingElement);
+    }
+
+    const movingIds = selectedRows.map(row => row.className.split(' ')[0]);
```
Comment (josaphatim) [inline modules/core/site.js:1891]: There is an issue with this line. Try to drap and release the click you will see the issue.
Uncaught TypeError: Cannot read properties of undefined (reading 'className')

**Comment by josaphatim** [comment]
@jacob-js I noticed that dragging to **manage folders** link also starts the ajax request which shouldn't. Also the row moved is deleted but I think the current page should be reloaded instead.

**Comment by marclaporte** [comment]
@jacob-js Does it work with a touch interface (tablet or phone)

**Comment by jacob-js** [comment]
> @jacob-js Does it work with a touch interface (tablet or phone)

While it ideally should, the app interface currently lacks the capability to display both the message list and folders simultaneously on mobile and tablet devices. Consequently, simulating the drag-and-drop functionality on these devices isn't feasible.

**Comment by jacob-js** [comment]
> @jacob-js I noticed that dragging to **manage folders** link also starts the ajax request which shouldn't. Also the row moved is deleted but I think the current page should be reloaded instead.

@josaphatim

1. I'm not sure if I got you clearly, are you suggesting that the messages should move between folders without triggering any AJAX request? If so, how could you do that?
2. Yes of course the row moved is deleted in the interface, I certainly agree that we could reload the page and fetch the entire list of messages, but this seems to me inappropriate because we just need to reflect the changes on the list without disrupting the user from what they are doing.

File: modules/core/site.js (hunk 1 above)
Comment (jacob-js) [reply modules/core/site.js:1897]: It seems to be impossible for me to reproduce this error. Any specific indication?

**Comment by jacob-js** [comment]
> @jacob-js I noticed that dragging to **manage folders** link also starts the ajax request which shouldn't. Also the row moved is deleted but I think the current page should be reloaded instead.

@josaphatim 

I got it, I added a check to allow only valid folders.

**Comment by jacob-js** [comment]
> > @jacob-js Does it work with a touch interface (tablet or phone)
> 
> While it ideally should, the app interface currently lacks the capability to display both the message list and folders simultaneously on mobile and tablet devices. Consequently, simulating the drag-and-drop functionality on these devices isn't feasible.

@marclaporte here is what I'm talking about in visual.
image
image

**Comment by marclaporte** [comment]
> > > @jacob-js Does it work with a touch interface (tablet or phone)
> > 
> > 
> > While it ideally should, the app interface currently lacks the capability to display both the message list and folders simultaneously on mobile and tablet devices. Consequently, simulating the drag-and-drop functionality on these devices isn't feasible.
> 
> @marclaporte here is what I'm talking about in visual.


Yes, I understand. Some laptops have a touch screen, but it's not so common.

**Comment by josaphatim** [comment]
Now the whole is working as expected, no error messages in the console.
But can you fix this test please: test_page_js_debug ?

**Pull request merged**